server/routes/bookings/cancelVisitController.ts

```diff
@@ -0,0 +1,112 @@
+import type { RequestHandler } from 'express'
+import { Meta, ValidationChain, matchedData, param, body, validationResult } from 'express-validator'
+import { BookerService, VisitService } from '../../services'
+import paths from '../../constants/paths'
+
+export default class CancelVisitController {
+  public constructor(
+    private readonly bookerService: BookerService,
+    private readonly visitService: VisitService,
+  ) {}
+
+  public confirmCancelView(): RequestHandler {
+    return async (req, res) => {
+      const { booker, bookings } = req.session
+      const { visits } = bookings
+
+      const errors = validationResult(req)
+      if (!errors.isEmpty() || bookings.type !== 'future') {
+        return res.redirect(paths.BOOKINGS.HOME)
+      }
+
+      const prisoner = booker.prisoners?.[0]
+        ? booker.prisoners?.[0]
+        : (await this.bookerService.getPrisoners(booker.reference))?.[0]
+
+      const { visitDisplayId } = matchedData<{ visitDisplayId: string }>(req)
+
+      const visit = visits.find(v => v.visitDisplayId === visitDisplayId)
+
+      return res.render('pages/bookings/cancel/cancel', {
+        errors: req.flash('errors'),
+        booker,
+        prisoner,
+        visit,
+        visitDisplayId,
+        showServiceNav: true,
+      })
+    }
+  }
+
+  public submit(): RequestHandler {
+    return async (req, res, next) => {
+      const { booker, bookings } = req.session
+      const { visits } = bookings
+      const { cancelBooking } = req.body
+
+      const { visitDisplayId } = matchedData<{ visitDisplayId: string }>(req)
```

Comment:
```suggestion
      const { cancelBooking, visitDisplayId } = matchedData<{ cancelBooking: 'yes' | 'no', visitDisplayId: string }>(req)
```

Reply:
is this not replacing the functionality introduced by the validation?

Reply:
No, it's typing it. So, if you check the type of `cancelBooking` having got it straight from `req.body` it'll be something like `any`. In reality, the actual *value* could be anything - because invalid data will be present on `req.body`. (By using `matchedData` with an expected type passed in, `cancelBooking` will have the type `'yes '| 'no'` (see docs [here](https://express-validator.github.io/docs/api/matched-data/#typescript-usage)).

We can be sure it is actually this (and not optional) because it wouldn't appear in `matchedData` if it was empty or an unexpected value.